docs/spec_draft.md

```diff
@@ -341,6 +351,175 @@ only exists to establish data dependencies from `result` to `inputs`.
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.all_reduce
+
+### Semantics
+
+For each replica sub-group in `replica_groups`, the operation performs
+[reduction](#stablehloreduce), using `computations`, over the replicas of
+`input` tensor formed using the subgroup and produces a `result` tensor.
+
+`channel_handle`, an optional input, is a handle given to a user to represent a
+communication channel between two computations via a
+[stablehlo.send](#stablehlosend) and [stablehlo.recv](#stablehlorecv)
+instruction pair or via collective instructions (e.g.
+[stablehlo.all_reduce](#stablehloall_reduce),
+[stablehlo.all_gather](#stablehloall_gather), etc.). It is represented using a
+`channel-id` and a `channel-type`. A `channel-id` is used for cross-partition
+communication; only operations with the same opcode and `channel-id` can
+communicate to each other. Non-positive `channel-id` is equivalent to no channel
+id. The `channel-type` is not relevant for this instruction.
+
+In the instruction syntax, multiple replica ids are grouped as a replica
+subgroup and multiple replica subgroups form `replica_groups`. If
+`use_global_device_ids = true`, the ids in `replica_groups` are interpreted as
+global device ids given by `rid * num_pids + pid` where `num_pids` $=$ number of
+partitions.
+
+The execution of the operation involves logically partitioning each execution
+instance (`ei`) of `rid`, `pid` pair, henceforth referred to as `ei(rid, pid)`
+into groups. All the execution instances within a group can participate in the
+operation without getting interfered by execution instances from other groups.
+The number of groups represents the number of `stablehlo.all_reduce` operations
+executed in parallel.
+
+Given,
+  * `num_pids`: Number of partition ids
+  * `num_rids`: Number of replica ids
+  * `all_partions_ids`: { pid : pid $\in$ [0, `num_pids`) }
+  * `all_replica_ids`: { rid : rid $\in$ [0, `num_rids`) }
+  * `size_subgroup`: size of each subgroup in `replica_groups`.
+  * `num_subgroup`: number of subgroups in `replica_groups`.
+
+The formation of groups is defined as follows:
+
+1. If `channel_handle` is not provided and `use_global_device_ids` $=$ `false`
+   * Number of groups formed: `num_subgroup` * `num_pids`.
+   * Size of each group: `size_subgroup`.
+   * The formation of `groups` can be demonstrated, using Python-like syntax, as
+     follows:
+     ```python
+     all_replica_ids = [rid for rid in range(num_rids)]
+     all_partition_ids = [pid for pid in range(num_pids)]
+     groups = []
+
+     if len(replica_groups) == 0:
+       replica_groups = [all_replica_ids]
+
+     for replica_group in replica_groups:
+       for pid in all_partition_ids:
+         sub_group = []
+         for rid in replica_group:
+           sub_group.append(ei(rid, pid))
+         groups.append(sub_group)
```

Comment:
```python
groups.append([ei(rid, pid) for rid in replica_group])
```

Reply:
All these pros will be simplified based on https://github.com/openxla/stablehlo/pull/634